### PR TITLE
Upgrade "Partner Several Companies" Module.

### PR DESCRIPTION
**Pull request by richard-willdooit**
OCA module upgrade to version 9.0.

### CODE REVIEW

File: partner_contact_in_several_companies/README.rst [hunk 1]
```diff
@@ -1,8 +1,10 @@
...
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
 
-Module name
-===========
+====================================
+Partner Contact in Several Companies
+====================================
 
 This module was written to extend the contact management functionality. It
```
Comment (elicoidal) [inline partner_contact_in_several_companies/README.rst]: This module extends

File: partner_contact_in_several_companies/README.rst [hunk 2]
```diff
@@ -30,10 +30,26 @@ For further information, please visit:
...
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/
+partner-contact/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback `here <https://github.com/OCA/
+partner-contact/issues/new?body=module:%20
+partner_contact_in_serverl_companies%0Aversion:%20
```
Comment (elicoidal) [inline partner_contact_in_several_companies/README.rst]: mispell

File: partner_contact_in_several_companies/README.rst [hunk 3]
```diff
@@ -42,12 +58,13 @@ Contributors
...
-* EL HADJI DEM <elhadji.dem@savoirfairelinux.com>
+* El Hadji Dem <elhadji.dem@savoirfairelinux.com>
 * TheCloneMaster <the.clone.master@gmail.com>
 * Sandy Carter <bwrsandman@gmail.com>
 * Rudolf Schnapka <rs@techno-flex.de>
 * Sebastien Alix <sebastien.alix@osiell.com>
 * Jairo Llopis <j.llopis@grupoesoc.es>
+* Richard deMeester <richard@willowit.com.au>
 
 Maintainer
```
Comment (elicoidal) [inline partner_contact_in_several_companies/README.rst:69]: Please use standard description

File: partner_contact_in_several_companies/__init__.py [hunk 4]
```diff
@@ -1,19 +1,4 @@
 # -*- coding: utf-8 -*-
```
Comment (elicoidal) [inline partner_contact_in_several_companies/__init__.py:1]: Please use simple headers following OCA templates
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

File: partner_contact_in_several_companies/__openerp__.py [hunk 5]
```diff
@@ -1,28 +1,27 @@
...
 
 {
     "name": "Contacts in several partners",
     "summary": "Allow to have one contact in several partners",
-    "version": "8.0.1.0.0",
-    "author": "Odoo Community Association (OCA)",
+    "version": "9.0.1.0.0",
     "category": "Customer Relationship Management",
     "website": "https://odoo-community.org/",
+    "author": "Odoo Community Association (OCA)",
```
Comment (elicoidal) [inline partner_contact_in_several_companies/__openerp__.py:10]: add main authors

File: partner_contact_in_several_companies/demo/res_partner.xml [hunk 6]
```diff
@@ -1,29 +1,27 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<openerp>
-    <data>
+<odoo>
```
Comment (elicoidal) [inline partner_contact_in_several_companies/demo/res_partner.xml]: remove empty files

File: partner_contact_in_several_companies/models.py [hunk 7]
```diff
@@ -1,232 +0,0 @@
...
-#    but WITHOUT ANY WARRANTY; without even the implied warranty of
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Affero General Public License for more details.
-#
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
-
-from openerp.osv import fields, orm, expression
```
Comment (elicoidal) [inline partner_contact_in_several_companies/models.py]: alphabetical sort

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
+# -*- coding: utf-8 -*-
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import fields, models, _, api
```
Comment (elicoidal) [inline partner_contact_in_several_companies/models/multi_contact.py]: alphabetical sort

File: partner_contact_in_several_companies/tests/test_partner_contact_in_several_companies.py [hunk 9]
```diff
@@ -127,8 +137,8 @@ def test_03_search_match_attached_contacts(self):
         cr, uid = self.cr, self.uid
         # Bob's contact has one other position which is related to
         # 'YourCompany'
-        # so search for all contacts working for 'YourCompany' should contain
-        # bob position.
+        # so search for all contacts working for 'YourCompany'
+        # should contain bob position.
```
Comment (elicoidal) [inline partner_contact_in_several_companies/tests/test_partner_contact_in_several_companies.py]: Capital

File: partner_contact_in_several_companies/README.rst (hunk 3 above)
Comment (richard-willdooit) [inline partner_contact_in_several_companies/README.rst:69]: Sorry, not sure what you mean, here?

File: partner_contact_in_several_companies/__init__.py (hunk 4 above)
Comment (richard-willdooit) [inline partner_contact_in_several_companies/__init__.py:1]: Is this not the correct header?

Comment (elicoidal) [inline partner_contact_in_several_companies/__init__.py:1]: You are right, I didnot see the second line.
Forget it

File: partner_contact_in_several_companies/README.rst (hunk 3 above)
Comment (richard-willdooit) [inline partner_contact_in_several_companies/README.rst:69]: I cannot find where this differs from the template.  I am sorry.

File: partner_contact_in_several_companies/README.rst (hunk 1 above)
Comment (richard-willdooit) [inline partner_contact_in_several_companies/README.rst]: I agree that wording sounds better.  Perhaps the template at https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst should be modified as the wording was copied from there.

File: partner_contact_in_several_companies/README.rst (hunk 3 above)
Comment (pedrobaeza) [inline partner_contact_in_several_companies/README.rst:69]: Nothing, it's correct, so go ahead. Sorry for the noise

File: partner_contact_in_several_companies/__openerp__.py (hunk 5 above)
Comment (pedrobaeza) [inline partner_contact_in_several_companies/__openerp__.py:10]: In authors: company names. In contributors (in README and here): people names

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
+# -*- coding: utf-8 -*-
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: Make one file per model, naming the file as the model. As example, res_partner.py. The only exception when there's a submodel, like sale.order.line, that is also included in the main model file (sale_order.py).

File: partner_contact_in_several_companies/README.rst (hunk 3 above)
Comment (elicoidal) [inline partner_contact_in_several_companies/README.rst]: SorryI got tricked by github...
Folded content :/

Comment (elicoidal) [inline partner_contact_in_several_companies/README.rst:69]: SorryI got tricked by github...
Folded content :/

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import fields, models, _, api
+from openerp.osv import expression
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    contact_type = fields.Selection([('standalone', _('Standalone Contact')),
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: Move the first argument line to the second line, and thus you only need to indent the line one level, being less constrained for putting all the text:

```
contact_type = fields.Selection(
    [('standalone', _('Standalone Contact')),
    ...
```

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+                                    required=True, select=1, store=True)
+    contact_id = fields.Many2one('res.partner', string='Main Contact',
+                                 domain=[('is_company', '=', False),
+                                         ('contact_type', '=', 'standalone'),
+                                         ],
+                                 )
+    other_contact_ids = fields.One2many('res.partner', 'contact_id',
+                                        string='Others Positions')
+
+    @api.one
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: @api.one is deprecated in v9. Use @api.multi

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    @api.depends('contact_id')
+    def _get_contact_type(self):
+        self.contact_type = self.contact_id and 'attached' or 'standalone'
+
+    _defaults = {
+        'contact_type': 'standalone',
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: Use `default` attribute directly in the field definition.

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    def _basecontact_check_context(self, mode):
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: @api.model?

File: partner_contact_in_several_companies/README.rst (hunk 1 above)
Comment (elicoidal) [inline partner_contact_in_several_companies/README.rst]: ACtually we have modified it lately

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+    @api.model
+    def create(self, vals):
+        modified_self = self._basecontact_check_context('create')
+        if not vals.get('name') and vals.get('contact_id'):
+            vals['name'] = modified_self.browse(vals['contact_id']).name
+        return super(ResPartner, modified_self).create(vals)
+
+    @api.multi
+    def read(self, fields=None, load='_classic_read'):
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: For what is this needed?

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+        modified_self = self._basecontact_check_context('write')
+        return super(ResPartner, modified_self).write(vals)
+
+    @api.multi
+    def unlink(self):
+        modified_self = self._basecontact_check_context('unlink')
+        return super(ResPartner, modified_self).unlink()
+
+    @api.multi
+    def _commercial_partner_compute(self, name, args):
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: This should be converted to new style computed field.

File: partner_contact_in_several_companies/models/multi_contact.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+            return
+        contact_fields = self._contact_fields()
+        contact_vals = dict(
+            (field, vals[field]) for field in contact_fields if field in vals
+        )
+        if contact_vals:
+            self.with_context(__update_contact_lock=True).write(contact_vals)
+
+    @api.model
+    def _fields_sync(self, partner, update_values):
```
Comment (pedrobaeza) [inline partner_contact_in_several_companies/models/multi_contact.py]: Why don't just simply add the fields to commercial fields?

Comment (richard-willdooit) [inline partner_contact_in_several_companies/models/multi_contact.py]: I don't know - it is not my module.  I just want to use it version 9, so I upgraded it.  I just need a working version accepted.

**Comment by richard-willdooit** [comment]
@pedrobaeza @elicoidal - I do not at all know how to interpret the "coverage/coveralls" and why it has a cross, now - can you either tell me what was wrong or how to interpret it?

**Comment by pedrobaeza** [comment]
This is because this PR increases the code quantity that is not tested.

**Comment by richard-willdooit** [comment]
So, how do we get pass this? All I did, in reality, was split the code in to two py files.

I really really really want to get this resolved and merged - it seems to be taking a very long time, and I did it because we wanted to use the code.

I presume the runbot fails because the coverage has failed?

**Comment by elicoidal** [comment]
@richard-willowit I am no technical expert but maybe you can add some simple test where it is missing. Coverall should be able to indicate which piece of code is not covered by the current test set.
Besides, maybe you should check the warning on runbot about the birthdate_date field?

**Comment by elicoidal** [comment]
If you check coverall:
screenshot from 2015-11-30 08-29-53

check ir_action.py coverage

**Comment by richard-willdooit** [comment]
I have written a test for ir_actions, and it is still failing - I have no idea what more I can do at this point!

I really just want to get this merged - it is such a straightforward change - is there anyone who can help me? I don't even know what the coverage/coveralls messages means!

**Comment by elicoidal** [comment]
@richard-willowit As far as I am concerned LGTM
cc @pedrobaeza Travis is green and coverage high: can we merge it?

**Comment by pedrobaeza** [comment]
Just right now the coverage has decreased, but it can be enough.

One last thing. Please fix this warning in runbot:

```
2015-12-01 04:01:09,369 21303 WARNING 3136340-197-57078a-all openerp.models: res.partner.create() includes unknown fields: birthdate_date
```

**Comment by richard-willdooit** [comment]
So, are we good to go?

**Comment by elicoidal** [comment]
LGTM cc @pedrobaeza

**Comment by pedrobaeza** [comment]
:+1:

**Pull request merged**